app/preload/database/transactionsDao.ts

```
@@ -161,45 +162,97 @@ async function hasPendingShieldedBalanceTransfer(fromAddress: string) {
+        .select();
+
+    const [alsoUpdates, additions] = partition(additionsOnHash, (t) =>
+        existingTransactions.some((t_) => t.id === t_.id)
     );
 
+    return { updates: updates.concat(alsoUpdates), additions };
+}
+
+/**
+ * Upserts the provided list of transactions into the database as part of a database transaction.
+ * @param updates the array of transactions to update
+ * @param additions the array of transactions to insert
+ * @param trx knex transaction, which these upserts should be added to.
+ */
+async function upsertTransactionsTransactionally<
+    T extends Partial<TransferTransaction>
+>(updates: T[], additions: T[], trx: Knex.Transaction): Promise<void> {
+    const additionChunks = chunkArray(additions, 50);
+
+    for (const additionChunk of additionChunks) {
+        await trx.table(transactionTable).insert(additionChunk);
+    }
+
+    for (const updatedTransaction of updates) {
+        const { transactionHash, ...otherFields } = updatedTransaction;
+        if (transactionHash) {
+            await trx
+                .table(transactionTable)
+                .where({ transactionHash })
+                .update(otherFields);
+        } else {
+            await trx
+                .table(transactionTable)
+                .where({ id: updatedTransaction.id })
+                .update(otherFields);
+        }
+    }
+}
+
+/** Given a list of transactions, checks which already exists.
+ *  New transactions are added to the table, while duplicates are treated
+ *  as updates to the current transactions.
+ * @Return the list of new transactions.
+ * */
+export async function insertTransactions(
+    transactions: Partial<TransferTransaction>[]
+) {
+    const { updates, additions } = await findExistingTransactions(transactions);
+    await (await knex())
+        .transaction(async (trx) => {
+            await upsertTransactionsTransactionally(updates, additions, trx);
+        })
+        .catch((e) => {
```

To me it seems you're kind of mixing the styles suggested in the knex documentation. For the async/await case they enclose the transaction in a try/catch block instead of having this re-throw.

---

I have removed this re-throwing. I added it because I, mistakenly, thought that knex was suppressing the error otherwise.